### ISSUE TITLE
using run_watch blocks indefinitely

### ISSUE THREAD

**Issue by or-adar** [devspace-sh/devspace#2565]
Hi
I am looking into using devspace features for my development process.
I have a sidecar written in go which I want to run in remote debugging and I want to allow hot reloading,
so if files are getting changed the container should be restarted and the binary gets built again.

I was looking into the documentation which suggested using "run_watch" to reload a container if certain conditions happen,
and I reached to the following devspace.yaml file:

```
version: v2beta1

name: my-service

commands:
  watch: |-
    run_watch -p devspace.yaml -p **/*.go -e **/*_test.go -- devspace dev "$@"

hooks:
  - name: go-build
    command: |-
      GOARCH=386 GOOS=linux go build -gcflags "all=-N -l"
    events: ["devCommand:before:execute"]

dev:
  app:
    container: manager
    imageSelector: my-image
    command:
      - bash
    args:
      - -c
      - >-
        while [ ! -f "/v1" ] || [ ! -f "/dlv" ]; do echo "syncing /v1 and /dlv"; sleep 15s; done &&
        echo 'done syncing' &&
        chmod +x /v1 &&
        chmod +x /dlv &&
        /dlv --listen=:2345 --headless=true --api-version=2 --accept-multiclient --log exec /v1
    sync:
      - path: ./v1:/v1
        file: true
        onUpload:
          restartContainer: true
      - path: "${DEVSPACE_USER_HOME}/go/bin/linux_386/dlv:/dlv"
        file: true
        onUpload:
          restartContainer: true
    ports:
      - port: 2345:2345
```

I'm doing some hacks here as I am not very familiar with devspace capabilities
so if you have any suggestions I'll be glad to here!
but either way, with the yaml file above, when I run "devspace run watch" the command is stuck indefinitely and don't see any helpful output, and if I may note, when I use more specific flags with the "run_watch" blocks things are fine but devspace is missing ".go" files in nested directories. for example:
if I put: `run_watch -p devspace.yaml -p *.go -- devspace dev "$@"` running "devspace run watch" emits no errors and runs as expected, however if I make change to a nested file "./nested_dir/nested_file.go" and save devsapce doesn't pick it up.
So, am I missing something here? should I be using something else to restart a container on nested file changes?

Also, is there any more appropriate way to change the file modes of synced files?
you can see in my example that I'm overriding the entrypoint of the image and run "chmod" to make certain files executable and I would like to hear whether you have other suggestions to accomplish this.

**Comment by lizardruss** [comment]
Hello! Thanks for reaching out.

I believe to see any output, you would need to add `logs` or `terminal` to the `dev.app` section:
```yaml
dev:
  app:
    # existing configuration
    logs: {}
```

As-is, you've only configured port forwarding and file syncing which run in the background.

As for the file permissions, I'm guessing that `v1` and `dlv` are binaries copied from your workspace. I think in that case `chmod` is necessary.

A slightly different approach might be to change your command to run against the source and have it build in the container. An example might be:
```yaml
dev:
  app:
    container: manager
    imageSelector: my-image
    devImage: loftsh/go
    command:
      - bash
    args:
      - -c
      - >-
        while [ ! -f "/v1" ] || [ ! -f "/dlv" ]; do echo "syncing /v1"; sleep 15s; done &&
        echo 'done syncing' &&
        dlv --listen=:2345 --headless=true --api-version=2 --accept-multiclient --log debug /v1/main.go
    sync:
      - path: ./:/v1
        onUpload:
          restartContainer: true
    ports:
      - port: 2345:2345
```

If you use a `devImage` with `dlv` already installed, you can skip the syncing of the `dlv` binary. This is based off our example project [here](https://github.com/loft-sh/devspace-example-go/tree/main/04-debugging#delve-dlv), which doesn't use `run_watch` in favor of running `dlv` from the terminal, but could be adapted to use `run_watch`.

**Comment by lizardruss** [comment]
Hello! Thanks for reaching out.

I believe to see any output, you would need to add `logs` or `terminal` to the `dev.app` section:
```yaml
dev:
  app:
    # existing configuration
    logs: {}
```

As-is, you've only configured port forwarding and file syncing which run in the background. `logs` should show the output from the command you've configured.

As for the file permissions, since `v1` and `dlv` are binaries copied from your workspace, the `chmod` call is needed. You could try using the `exec` option to chmod them after uploading, for example:
```yaml
    sync:
      - path: ./v1:/v1
        file: true
        onUpload:
          restartContainer: true
          exec:
          - command: |-
              chmod +x /v1
      - path: "${DEVSPACE_USER_HOME}/go/bin/linux_386/dlv:/dlv"
        file: true
        onUpload:
          restartContainer: true 
          exec:
          - command: |-
              chmod +x /dlv
```

However, because of the container restart, it may be necessary to also configure [persistent paths](https://www.devspace.sh/docs/configuration/dev/modifications/persistence). Otherwise the container will restart and erase the file mode change.

FWIW, the more common approach is to sync the source code to the container and build it there. Using a [devImage](https://www.devspace.sh/docs/configuration/dev/modifications/dev-image) makes it so `dlv` is already installed. An example project is [here](https://github.com/loft-sh/devspace-example-go), but it uses a different workflow from what you're aiming for.

I think your approach could still work. Here's a configuration based on yours that worked for me:
```yaml
version: v2beta1

name: my-service

# Can remove, for example purposes only
deployments:
  app:
    helm:
      values:
        containers:
          - image: nginx # replace with my-image

dev:
  app:
    container: manager
    imageSelector: nginx # replace with my-image
    devImage: loftsh/go
    command:
      - dlv
    args:
      - debug
      - /app/main.go
      - --listen=0.0.0.0:2345
      - --headless
      - --api-version=2
      - --output=/tmp/__debug_bin
    sync:
      - path: .:/app
        onUpload:
          restartContainer: true
    ports:
      - port: 2345:2345
    logs: {}
```

In this case, `dlv` already exists in `loftsh/go` and `dlv` is compiling the application from `/app/main.go`, so the `chmod` step wasn't needed. With `logs` config added, you should see output similar to the following after running `devspace dev`:
```sh
dev:app sync  Initial sync completed
dev:app logs  Streaming logs of pod:container dep-chart-devspace-645fcbbdbf-qj55v:dep-chart
dev:app logs  Container started with restart helper.
dev:app logs  Waiting for initial sync to complete or file /.devspace/start to exist before starting the application...
dev:app logs  Starting application...
dev:app logs  API server listening at: [::]:2345
dev:app logs  2023-02-22T18:41:42Z warning layer=rpc Listening for remote connections (connections are not authenticated nor encrypted)
```

Connect your debugger and things should work. I hope this was helpful!

**Comment by or-adar** [comment]
Hi @lizardruss 
Appreciate your suggestions! I'll look into them!
thought about using sync for the entire source code to a container
but my issue was that I have private packages,
however, I think that I can mount (using sync) any installed packages from "$GOPATH/pkg" to the running container.
but what I don't get is how can you run dlv as the entrypoint? 
(assuming my devImage is going to be the official golang image and not loftsh/go)
Isn't it like the "chicken or the egg" paradox?
devspace performs the initial sync when the pod is ready to upload main.go (and other files)
but for the pod to be ready the entrypoint command should be successful (and dlv relies on the existence of main.go which is uploaded only at a later phase).

the only solution I can think of is to have a custom image with a fake "main.go" for the first time the container is created (until it gets overriden when "sync" kicks in) but I don't like the idea of managing some additional images and dockerfiles..

**Comment by lizardruss** [comment]
@or-adar Good question! When you use `restartContainer` DevSpace overrides the container with a restart helper script that will wait for the initial sync to complete before running the command supplied from `devspace.yaml`.

For private packages, you may have luck with [proxyCommands](https://www.devspace.sh/docs/configuration/dev/connections/proxy-commands). I haven't tried this myself, but it may be worth a try.. add this to the `dev.app` section, and see if private packages work:
```yaml
    proxyCommands:
      - gitCredentials: true
      - command: go
```

**Comment by lizardruss** [comment]
@or-adar Did you arrive at a configuration that works for you? If not, feel free to re-open this issue and we'll take a look!
